src/blas/impl/KokkosBlas_Newton_impl.hpp

```diff
@@ -0,0 +1,250 @@
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Luc Berger-Vergiat (lberge@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef __KOKKOSBATCHED_ODE_NEWTON_HPP__
+#define __KOKKOSBATCHED_ODE_NEWTON_HPP__
+
+#include "Kokkos_Core.hpp"
+#include "KokkosBatched_LU_Decl.hpp"
+#include "KokkosBatched_LU_Serial_Impl.hpp"
+#include "KokkosBatched_Scale_Decl.hpp"
+#include "KokkosBatched_Gesv.hpp"
+#include "KokkosBlas1_nrm2.hpp"
+#include "KokkosBlas1_scal.hpp"
+#include "KokkosBlas1_axpby.hpp"
+
+namespace KokkosBlas {
+namespace Impl {
+
+enum class NewtonSolverStatus { Converged = 0, LinearSolveFailure, MaxIters };
+
+std::ostream& operator<<(std::ostream& os, NewtonSolverStatus& status) {
+  switch (status) {
+    case NewtonSolverStatus::Converged: os << "Newton Solver Converged!"; break;
+    case NewtonSolverStatus::LinearSolveFailure:
+      os << "Newton: Linear Solver Failure";
+      break;
+    case NewtonSolverStatus::MaxIters:
+      os << "Newton reached maximum iterations without convergence.";
+      break;
+  }
+  return os;
+}
+
+/// \brief NewtonHandle
+///
+/// This handle is used to pass information between the Newton Solver and
+/// the calling code.
+///
+/// \tparam: NormViewType: Type of view used to store the residual convergence
+/// history
+
+template <class NormViewType>
+struct NewtonHandle {
+  using norm_type = typename NormViewType::non_const_value_type;
+
+  NormViewType lastResidual;  // Residual of last successful iteration
+  typename NormViewType::HostMirror lastResidualHost;
+
+  // NormViewType  residual_norms;
+  // TODO: Making these public for now. Should make private and access
+  // via setters and getters?
+  int maxIters;           // Maximum number of Newton steps
+  norm_type relativeTol;  // Relative convergence tolerance
+  bool debug_mode;        // Returns extra verbose output if true.
+
+  NewtonHandle(int _maxIters = 25, double _relativeTol = 1.0e-6,
+               bool _debug = false)
+      : lastResidual("ending Residual norm", 1),
+        lastResidualHost("end res norm host", 1),
+        maxIters(_maxIters),
+        relativeTol(_relativeTol),
+        debug_mode(_debug) {}
+
+  norm_type get_end_residual() const {
+    Kokkos::deep_copy(lastResidualHost, lastResidual);
+    return lastResidualHost(0);
+  }
+
+};  // NewtonHandle
+
+/// \brief Newton Functor:
+/// Solves the nonlinear system F(x) = 0
+/// where F is a map from R^n to R^m.
+/// \tparam LHSFunc: Functor on (J,x) where matrix J is output with
+/// the value of the Jacobian evaluated at x.
+/// \tparam RHSFunc: Functor on (y,x) where y is output with the value
+/// of F at x.
```

Comment:
This is no more consistent with the structure as now you use a `System ` class.

Reply:
Changed